### DISCUSSION TITLE
About fastapi deployment

### DISCUSSION THREAD

**Discussion by Zengyf-CVer** [fastapi/fastapi#8762] [Questions] [answered]
@tiangolo 

Hello, I have used your fastapi framework very well, and I have done some related projects. But I have a question, how do I deploy the fastapi project to my server, such as nginx?

**Comment by panla** [comment]
example

use docker or use supervisor

```bash
uvicorn main:app
uvicorn main:app --host 0.0.0.0 --port 8000
uvicorn main:app --host 0.0.0.0 --port 8000 --http httptools --loop uvloop
uvicorn main:app --host 0.0.0.0 --port 8000 --http httptools --loop uvloop --workers 4
```

```config
server {
    listen 80;
    server_name example.com;

    location / {
        proxy_set_header Host $http_host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_pass http://127.0.0.1:8000;
    }
}
```

**Comment by Zengyf-CVer** [comment]
@panla 

Hello, first of all thank you for your reply, but I have two questions:
- Do I need to install nginx directly in this machine or install nginx in docker?
- Due to conditions, I can only install nginx in the ubuntu 20.04 system. Will this affect the deployment of fastapi?

**Comment by panla** [comment]
**all fine**

## example

### There is a docker network

study docker network

```bash
docker network create --driver bridge --subnet 172.18.0.0/16 --gateway 172.18.0.1 dev_net
```

```text
api server, host='0.0.0.0' port=8000  docker network is 172.18.0.2

whether nginx`s is in host machine or container

no port map:
    nginx can proxy_pass http://172.18.0.2:8000
ther is a port map:
    nginx can proxy_pass http://172.18.0.2:8000
    nginx can proxy_pass http://127.0.0.1:9000
```

### no docker network and There is a public network

```text
api server, host='0.0.0.0' port=8000, docker port map is 127.0.0.1:9000:8000

whether nginx is in host machine or container

nginx can proxy_pass http://public_ip:9000, need allow this port
or
nginx can proxy_pass http://127.0.0.1:9000
```

### no docker network and no public network

```text
api server, host='0.0.0.0' port=8000, docker port map is 127.0.0.1:9000:8000

need install nginx in host machine

nginx need proxy_pass http://127.0.0.1:9000
```

**Comment by Zengyf-CVer** [comment]
@panla 
Hello, I would like to ask, how to deploy FastAPI to Alibaba Cloud ECS?

**Comment by panla** [comment]
### 方法一 docker

- 1, 宿主机安装 docker, nginx , docker-compose
- 2, 把你的项目用 docker/docker-compose 运行起来
- 3 用 ngin 转发至你服务的端口

### 方法二 supervisor

- 1 安装 ngixn, supervisor
- 2 把你的服务用 supervisor 运行起来
- 3 用 ngin 转发至你服务的端口

### 方法二  tmux

- 1 安装 ngixn, tmux
- 2 把你的服务在tmux里运行起来
- 3 用 nginx 转发至你服务的端口

### 不用 nginx

直接 uvicorn --host 0.0.0.0 然后服务器开放端口，也是可以的

**Comment by YuriiMotov** [comment, marked as answer]
See Docs: https://fastapi.tiangolo.com/deployment/
